### PR TITLE
enhancement(apt platform,debian platform,dpkg platform,centos platform,rhel platform,rpm platform,yum platform): Improve packaging and service best practices

### PR DESCRIPTION
**Pull request by spencergilbert**
Closes: https://github.com/vectordotdev/vector/issues/12552
Closes: https://github.com/vectordotdev/vector/issues/12551

The original intent of this work was to adhere to the [target spec](https://github.com/vectordotdev/vector/blob/master/docs/specs/target.md), though after discussion with other teams at Datadog we were leaning too far into secure by default and harming the UX and onboarding process.

The work here represents a general cleanup of the SystemD service, using more modern and (as far as I can tell) best practices. This pushed some changes to the linux packages (removing now unused files for example).

This also changes the default configuration for Vector to be an Aggregator (the old default did nothing but generate dummy logs into the system's journald ⛔ ), and provides an alternative Agent configuration available via systemd override configurations.

**EDIT**: Aug 7

Vector should no longer ship a default config, it doesn’t make sense given the tool it’s too flexible to provide a useful default, prefer to add complete and useful example configs - stop including “/etc/vector/vector.toml” in distributed packages.

- SystemD, no default - without args/opts Vector errors

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -37,13 +37,14 @@ debug = true
...
-conf-files = ["/etc/vector/vector.toml", "/etc/default/vector"]
+conf-files = ["/etc/vector/vector.toml"]
 assets = [
   ["target/release/vector", "/usr/bin/", "755"],
   ["config/vector.toml", "/etc/vector/vector.toml", "644"],
   ["config/examples/*", "/etc/vector/examples/", "644"],
+  ["config/agent/*", "/etc/vector/agent/", "644"],
+  ["config/aggregator/*", "/etc/vector/aggregator/", "644"],
   ["distribution/systemd/vector.service", "/lib/systemd/system/vector.service", "644"],
-  ["distribution/systemd/vector.default", "/etc/default/vector", "600"]
```
Comment (spencergilbert) [inline Cargo.toml:46]: SystemD is rather anti-env files, so I decided to remove our default usage of it. It can still be used via override conf files.

File: config/vector.toml [hunk 2]
```diff
@@ -12,33 +12,71 @@
...
-inputs = ["parse_logs"]
-encoding.codec = "json"
-
-# Vector's GraphQL API (disabled by default)
-# Uncomment to try it out with the `vector top` command or
-# in your browser at http://localhost:8686
-#[api]
-#enabled = true
-#address = "127.0.0.1:8686"
+# The directory used for persisting Vector state.
```
Comment (spencergilbert) [inline config/vector.toml:15]: Converting the `aggregator/vector.yaml` into TOML and moved it to the default location

File: config/vector.toml [hunk 3]
```diff
@@ -12,33 +12,71 @@
...
+data_dir = "/var/lib/vector"
+
+# Configure Vector's GraphQL API.
+[api]
+enabled = true
+address = "[::]:8686"
+playground = false
+
```
Comment (spencergilbert) [inline config/vector.toml:23]: I am thinking that for ease of "quickstart" having an `http` source may be nice even if it's not totally practical to have a default config for such a generic component.

Without it though, there's not an super simple way to get an event into Vector in an Aggregator role (`curl` in an event).

File: distribution/rpm/vector.spec [hunk 4]
```diff
@@ -76,8 +82,9 @@ rm -rf %{buildroot}
 %{_bindir}/*
 %{_unitdir}/vector.service
 %config(noreplace) %{_sysconfdir}/%{_name}/vector.toml
-%config(noreplace) %{_sysconfdir}/default/vector
 %config %{_sysconfdir}/%{_name}/examples/*
+%config %{_sysconfdir}/%{_name}/agent/*
+%config %{_sysconfdir}/%{_name}/aggregator/*
```
Comment (spencergilbert) [inline distribution/rpm/vector.spec:87]: So this seems angry if it's just copying over when it's just a `.gitkeep` file, I'd like to keep the directory structure for future defaults.

I could copy the `config/vector.toml` into it but the duplication feels a little weird.

File: distribution/systemd/hardened-vector.service [hunk 5]
```diff
@@ -5,20 +5,24 @@
...
 
 [Service]
-EnvironmentFile=-/etc/default/vector
+Type=exec
 User=vector
 Group=vector
+# Use `systemctl edit vector` override environment variables,
+# or place `.conf` files under `/etc/systemd/system/vector.service.d/`.
+Environment="LOG_FORMAT=json"
+Environment="VECTOR_CONFIG_YAML=/etc/vector/aggregator/vector.yaml"
```
Comment (jszwedko) [inline distribution/systemd/hardened-vector.service:20]: Did you mean to add these to the non-hardened service file too? Or, rather, I don't understand why they are only in this one.

File: distribution/debian/scripts/postinst [hunk 6]
```diff
@@ -1,21 +1,49 @@
...
+  chmod 740 /var/lib/vector
+
+  # Own and RW restrict Vector config directory
+  chown -R vector:vector /etc/vector
+  chmod 740 /etc/vector
+}
+
+case "$1" in
+configure)
+  create_user
```
Comment (jszwedko) [inline distribution/debian/scripts/postinst:44]: This is making me realize we probably need a `postrm` script too to remove the user.

**Comment by spencergilbert** [comment]
Also remembered https://github.com/vectordotdev/vector/issues/13209 needs to be addressed - and it looks like we copy it to the same place for debs?

**Comment by spencergilbert** [comment]
> > Also remembered #13209 needs to be addressed - and it looks like we copy it to the same place for debs?
> 
> I think it's ok to leave that issue aside for now since it out-of-scope of the target spec. We can follow up on it.

Only a little scared of the package update process if we're moving things around. Would definitely like to get us CI around this first.

File: distribution/systemd/hardened-vector.service (hunk 5 above)
Comment (spencergilbert) [reply distribution/systemd/hardened-vector.service:20]: That's a whoopsie - copy pasted before I made changes to the non-hardened one. Fixed in 4d4e08553

File: config/vector.toml [hunk 7]
```diff
@@ -12,33 +12,71 @@
...
-#[api]
-#enabled = true
-#address = "127.0.0.1:8686"
+# The directory used for persisting Vector state.
+data_dir = "/var/lib/vector"
+
+# Configure Vector's GraphQL API.
+[api]
+enabled = true
+address = "[::]:8686"
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply config/vector.toml:21]: I _believe_ so, IIRC I tested in a k8s cluster that was ipv4 only.

https://serverfault.com/questions/21657/semantics-of-and-0-0-0-0-in-dual-stack-oses

File: distribution/debian/scripts/postinst (hunk 6 above)
Comment (spencergilbert) [reply distribution/debian/scripts/postinst:44]: https://wiki.debian.org/AccountHandlingInMaintainerScripts#:~:text=Account%20deletion,for%20example%20piuparts).

https://ma.ttias.be/on-removing-users-with-postun-in-rpm-spec-files/

This seems to be something up to debate, I opted for non-destructive.

File: distribution/systemd/hardened-vector.service [hunk 8]
```diff
@@ -5,20 +5,24 @@
...
-Description=Vector
-Documentation=https://vector.dev
+Description="Vector - An observability pipelines tool"
+Documentation=https://vector.dev/
+Wants=network-online.target
 After=network-online.target
-Requires=network-online.target
 
 [Service]
-EnvironmentFile=-/etc/default/vector
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply distribution/systemd/hardened-vector.service:14]: This was part of the best practices portion...

tl;dr SystemD now (always has?) preferred env vars to be defined in the conf files using `Environment`. These should be placed at `/etc/systemd/system/vector.service.d/foo.conf`

https://serverfault.com/questions/413397/how-to-set-environment-variable-in-systemd-service
https://askubuntu.com/questions/997818/location-to-place-systemd-service-environmentfile-when-creating-debian-package
https://www.freedesktop.org/software/systemd/man/systemd.service.html#Command%20lines

File: config/vector.toml [hunk 9]
```diff
@@ -12,33 +12,71 @@
 # Chat: https://chat.vector.dev
 # ------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply config/vector.toml:13]: Updated in: 8762da1a8257a54a6c776318f5ea6a9c143bc49c

File: distribution/debian/scripts/postinst (hunk 6 above)
Comment (jszwedko) [reply distribution/debian/scripts/postinst:44]: Hah, what mess. Yeah, I think removing the user on uninstall is out of scope anyway, so we can leave the existing behavior for now. I can see the arguments for both.

File: config/vector.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply config/vector.toml:25]: I am seeing this note on that thread:

> (except if you use Debian, which actually defaults to bindv6only = 1).

I wonder if we should just stick with ipv4 for now and follow up on the use of ipv6 later based on user feedback.

File: config/vector.toml (hunk 3 above)
Comment (jszwedko) [reply config/vector.toml:27]: 👍 I'd be down with an `http` source configured to accept ndjson.

File: distribution/systemd/hardened-vector.service (hunk 8 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply distribution/systemd/hardened-vector.service:14]: Mmm, yeah, I see. I'm open to that, but we need to document this as a breaking change. My preference, unless we have a super strong reason not to, would just be to maintain support for it. I can see the aesthetic argument from systemd maintainers, but given this is a compatibility break that is likely to affect a significant number of users, I think we should have a stronger one to make it.

File: config/aggregator/vector.toml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+type = "datadog_logs"
+inputs = [
+    "datadog_agent",
+    "fluent",
+    "logstash",
+    "splunk_hec",
+    "syslog",
+    "vector",
+]
+default_api_key = "${DATADOG_API_KEY}"
```
Comment (jszwedko) [inline config/aggregator/vector.toml:76]: I think we should just use the blackhole sink here as well and let users configure their preferred sinks. Making the default to send to Datadog harms Vector's position as a vendor-neutral observability pipeline.

File: config/agent/vector.toml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+# Configure Vector's GraphQL API.
+[api]
+enabled = true
+address = "[::]:8686"
+playground = false
+
+# Receives observability data from a Datadog Agent over HTTP or HTTPS.
+[sources.datadog_agent]
+type = "datadog_agent"
+address = "[::]:8282"
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply config/agent/vector.toml:27]: Gotcha, yeah, I see. I still think we should leave this source out of the default config as, similar to my comment in https://github.com/vectordotdev/vector/pull/13456/files#r916310631 I think this harms Vector's positioning. I realize there could be an argument made that the Datadog Agent is the only agent we integrate with that we don't recommend replacing, but until we have a second one, I think I'd rather just let users add it if they need.

File: distribution/systemd/hardened-vector.service (hunk 8 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply distribution/systemd/hardened-vector.service:14]: backwards compatibility- the Achilles Heel of improving software.

File: distribution/systemd/vector.service [hunk 12]
```diff
@@ -1,19 +1,25 @@
...
+#
+# To run Vector as an Agent with the default configuration
+# can set the following environment variable.
+# Environment="VECTOR_CONFIG=/etc/vector/agent/vector.toml"
 ExecStartPre=/usr/bin/vector validate
 ExecStart=/usr/bin/vector
 ExecReload=/usr/bin/vector validate
 ExecReload=/bin/kill -HUP $MAINPID
-Restart=no
+Restart=always
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply distribution/systemd/vector.service:21]: Seemed like an oversight tbh, Vector should always be a long running process - until/unless we add more batch processing options.

File: distribution/systemd/hardened-vector.service (hunk 8 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply distribution/systemd/hardened-vector.service:14]: I'll include the old way for compatibility and leave the comments on how to do it "systemd proper"

**Review by neuronull — APPROVED** [review]

File: distribution/debian/scripts/postinst [hunk 13]
```diff
@@ -1,21 +1,53 @@
...
+
+configure_permissions() {
+  if [ ! -d "/var/lib/vector" ]; then
+    # Create default Vector data directory
+    mkdir -p /var/lib/vector
+  fi
+
+  # Own and RW restrict Vector data directory
+  chown -R vector:vector /var/lib/vector
+  chmod 740 /var/lib/vector
```
Comment (prognant) [inline distribution/debian/scripts/postinst:35]: `install -g vector -o vector -m 740 /var/lib/vector` should work ?

Comment (spencergilbert) [reply distribution/debian/scripts/postinst:35]: That's spooky, never seen/used that particular command before - I'm ambivalent, happy to sub it in if that's what's preferred.

**Comment by spencergilbert** [comment]
@jszwedko what release do we want to get this out in `0.24` or `0.25`?

**Review by neuronull — APPROVED** [review]
